### ISSUE TITLE
Self-Implemented Metrics Behaved Weirdly in DDP Mode

### ISSUE THREAD

**Issue by donglihe-hub** [Lightning-AI/torchmetrics#2022]
## 🐛 Bug

Self-implemented nDCG@5 is larger than nDCG@1 in DDP mode. What I was trying to do here is to record the sum of sample-wise scores **_s_** and the number of samples **_n_**. At the end of batch/epoch, I can obtain the mean of nDCG@k by dividing **_s_** by **_n_**.

My implementation looks very similar to the example on [IMPLEMENTING A METRIC](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html). However, It failed in distributed settings.

`Epoch 8:  60%|██████    | 1541/2555 [04:58<03:16,  5.17it/s, loss=0.00354, v_num=213259, train_ndcg@1=0.656, train_ndcg@3=0.675, train_ndcg@5=0.698,...`

### To Reproduce

<details>
  <summary>Self-implemented Metric</summary>

```python
from torchmetrics.functional.retrieval import retrieval_normalized_dcg
from torchmetrics import Metric
import torch
from torch import Tensor

class NDCG(Metric):
    full_state_update = False

    def __init__(self, top_k: int):
        super().__init__()
        self.top_k = top_k
        self.add_state("ndcg", default=torch.tensor(0, dtype=torch.float), dist_reduce_fx="sum")
        self.add_state("n", default=torch.tensor(0), dist_reduce_fx="sum")

    def update(self, preds: Tensor, target: Tensor):
        # preds and target are 2-D tensors.
        assert preds.shape == target.shape

        self.ndcg += torch.stack([retrieval_normalized_dcg(p, t, k=self.top_k) for p, t in zip(preds, target)]).sum()
        self.n += len(preds)

    def compute(self):
        return self.ndcg / self.n
```

</details>

### Expected behavior

The documentation of self-implemented metric looked too simple and probably missed some details of distributed metrics.

I noticed that you used list and in `compute()` you call mean() to get the average score of a given batch. Can you please explain the difference between these two implementations?

### Environment
- TorchMetrics version (and how you installed TM, e.g. conda, pip, build from source):
0.10.3 (conda-forge)
- Python version:
3.10.12
PyTorch version (e.g., 1.0):
1.13.1
- Any other relevant information such as OS (e.g., Linux):
pytorch-lightning: 1.9.4
Linux: 5.15.0-78-generic

**Comment by SkafteNicki** [comment]
Hi @donglihe-hub, thanks for reporting this issue.
I investigated it a bit, but to me it seems that maybe your intuition behind nDCG@k is not correct. It is not impossible for nDCG@5 to be higher than nDCG@1, which is probably why you are seeing the results you are seeing.

Here is a example, which I computed both using torchmetrics and sklearn (just to confirm the behavior):
```python
import torch
from torchmetrics.functional.retrieval import retrieval_normalized_dcg
from sklearn.metrics import ndcg_score

target = torch.tensor([[1, 0, 0, 1, 0, 0, 1, 1, 1, 0]])
preds = torch.tensor([[0.54, 0.89, 0.50, 0.27, 0.33, 0.35, 0, 0.75, 0.53, 0.14]])

print([retrieval_normalized_dcg(preds, target, top_k=k) for k in [1,3,5]])
#prints [tensor(0.), tensor(0.5307), tensor(0.5296)]

print([ndcg_score(target, preds, k=k) for k in [1,3,5]])
#prints [0.0, 0.5307212739772434, 0.529634717214042]
```
that said I will be updating the documentation regarding implementing a metric in the near future.

**Comment by donglihe-hub** [comment]
@SkafteNicki Thank you for your reply. I think I might have some clues why this happened.

In the [IMPLEMENTING A METRIC](https://torchmetrics.readthedocs.io/en/stable/pages/implement.html) it said:

```
...
Similarly, calling compute() does the following internally:

1. Syncs metric states between processes.
...
```

I printed the number of samples and found that during training process the number is equal to the batch size on **a single GPU**, say 64. However, what I expect is 64 * 4 = 256 as I used 4 GPUs.

**Comment by donglihe-hub** [comment]
> ```python
> #prints [0.0, 0.5307212739772434, 0.529634717214042]
> ```

I tested my metric on Colab and obtained the same results.

```python
!pip install -q torchmetrics
import torchmetrics
from torchmetrics import Metric
from torchmetrics.functional.retrieval import retrieval_normalized_dcg

class NDCGnew(Metric):
    is_differentiable = False
    higher_is_better = True
    full_state_update = False

    def __init__(self, top_k: int):
        super().__init__()
        self.top_k = top_k
        # the range of ndcg is from 0 to 1
        self.add_state("ndcg", default=torch.tensor(0, dtype=torch.float), dist_reduce_fx="sum")
        self.add_state("n", default=torch.tensor(0), dist_reduce_fx="sum")

    def update(self, preds, target):
        assert preds.shape == target.shape
        self.ndcg += torch.stack([retrieval_normalized_dcg(p, t, top_k=self.top_k) for p, t in zip(preds, target)]).sum()
        self.n += len(preds)

    def compute(self):
        return self.ndcg / self.n

target = torch.tensor([[1, 0, 0, 1, 0, 0, 1, 1, 1, 0]])
preds = torch.tensor([[0.54, 0.89, 0.50, 0.27, 0.33, 0.35, 0, 0.75, 0.53, 0.14]])

m_1 = NDCGnew(1)
m_3 = NDCGnew(3)
m_5 = NDCGnew(5)
m_1.update(preds, target)
m_3.update(preds, target)
m_5.update(preds, target)
m_1.compute(), m_3.compute(), m_5.compute()
# (tensor(0.), tensor(0.5307), tensor(0.5296))
```

**Comment by donglihe-hub** [comment]
@SkafteNicki I forgot to thank you for correcting my understanding of nDCG in my last reply. Thank you for your help :)

As you have solved my issue, I will close it now. Thank you again for solving my problem! Hope you have a nice day!
